Runtime/Shaders/MapsBaker.cs

```diff
@@ -0,0 +1,238 @@
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+using UnityEngine;
+
+namespace Unity.MeshSync {
+/// <summary>
+/// Helper class to bake smoothness and metallic into channels required for the standard shaders.
+/// </summary>
+internal static class MapsBaker {
+    private const string _SmoothnessTextureChannel = "_SmoothnessTextureChannel";
+
+    static Dictionary<string, ComputeShaderHelper> loadedShaders = new Dictionary<string, ComputeShaderHelper>();
+
+    private const string SHADER_CONST_METALLIC   = "Metallic";
+    private const string SHADER_CONST_SMOOTHNESS = "Smoothness";
+    private const string SHADER_CONST_RGB        = "RGB";
+
+    private const string SHADER_FILE                       = "meshsync_channel_mapping";
+    private const string SHADER_NAME_SMOOTHNESS_INTO_ALPHA = "smoothness_into_alpha";
+    private const string SHADER_NAME_HDRP_MASK             = "hdrp_mask";
+
+
+    private static ComputeShaderHelper LoadShader(string name) {
+        string file = SHADER_FILE;
+
+        if (!loadedShaders.TryGetValue(file, out var shaderHelper)) {
+#if UNITY_EDITOR
+            var shaderFiles = AssetDatabase.FindAssets(file);
+            if (shaderFiles.Length > 0) {
+                var shaderFile = shaderFiles[0];
+                var shader = AssetDatabase.LoadAssetAtPath<ComputeShader>(AssetDatabase.GUIDToAssetPath(shaderFile));
+
+                shaderHelper = new ComputeShaderHelper(shader, name);
+
+                loadedShaders[name] = shaderHelper;
+            }
+#else
+// TODO: Load shader at runtime here
+#endif
+        }
+
+        return shaderHelper;
+    }
+
+    static MaterialPropertyData? FindMaterialProperty(List<MaterialPropertyData> materialProperties, string name) {
+        for (int i = 0; i < materialProperties.Count; i++) {
+            if (materialProperties[i].name == name) {
+                return materialProperties[i];
+            }
+        }
+
+        return null;
+    }
+
+    private static bool FindTexture(string name,
+        List<TextureHolder> textureHolders,
+        List<MaterialPropertyData> materialProperties,
+        string fallbackPropertyName,
+        float fallbackPropertyValue,
+        out Texture2DDisposable disposableTexture) {
+        var prop = FindMaterialProperty(materialProperties, name);
+
+        if (prop.HasValue) {
+            MaterialPropertyData.TextureRecord rec = prop.Value.textureValue;
+            disposableTexture = new Texture2DDisposable(BaseMeshSync.FindTexture(rec.id, textureHolders), false);
```

Comment:
The texture is only disposed if it was created on the fly, otherwise that is not necessary. Texture2DDisposable keeps track of that and disposes only when required.

Reply:
Keeping this comment open because it's related to another comment below in the same file.